source/agora/api/FullNode.d

```diff
@@ -260,6 +260,25 @@ public interface API
 
     public PreImageInfo getPreimage (Hash enroll_key);
 
+    /***************************************************************************
+
+        Params:
+            enroll_keys = set of enrollment keys
+
+        Returns:
+            The preimages for the specified enroll keys.
+            If the requested node doesn't know the preimage for a specific
+            enroll key, then it will not be included in the result.
+            If the set of enroll_keys is empty, then all preimages known to the
+            node are returned.
+
+        API:
+            GET /preimages_for_enroll_keys
+
+    ***************************************************************************/
+
+    public PreImageInfo[] getPreimagesForEnrollKeys (Set!Hash enroll_keys);
```

Comment:
1. there are a lot of convenient function on set setDifference that I use in many places to make the code more clear
2. followed this: getTransactions (Set!Hash tx_hashes); which I think the correct approach

Reply:
`setDifference` works on range, so it will work on arrays. Also, it assumes the input is sorted, which I don't think is the case here ?

Reply:
Regarding point 2, I don't *mind* having `Set` in the API. But:
- [x] It needs to be convenient to use from other languages: Check, because it serializes just like an array;
- [ ] It needs to be convenient to call from our own code: This is the current problem, because calling this with a single value is 3-4 lines of code;
- [x] It needs not to incur more penalty (speed, memory usage, etc...) than an array: I believe this is more-or-less the case right now (not exactly, because one can't manage the underlying memory of an AA, but it's a price we decided to pay), but it's something to keep in mind, especially if you solve the second point;

Reply:
'rename to getPreimages' - isn't that a bit ambiguous for a REST interface as we already have 

public override PreImageInfo[] getPreimages (ulong start_height,
        ulong end_height)
        
can Vibe.d even handle this (2 REST interfaces with the same endpoint name)?

Reply:
Vibe.d serializes Sets as dictionaries btw.

Reply:
>  Also, it assumes the input is sorted, which I don't think is the case here ?

'it assumes the input is sorted' - it is sorted in our current implementation

`It needs to be convenient to call from our own code: This is the current problem, because calling this with a single value is 3-4 lines of code;` 

extra memory allocation?